planner/core/find_best_task.go

```diff
@@ -936,26 +938,90 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
 	return ts, partialCost, nil
 }
 
-func indexCoveringCol(col *expression.Column, indexCols []*expression.Column, idxColLens []int) bool {
+func indexCoveringCol(col *expression.Column, constVal *expression.Constant, indexCols []*expression.Column, idxColLens []int) bool {
 	for i, indexCol := range indexCols {
-		isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
-		if indexCol != nil && col.Equal(nil, indexCol) && isFullLen {
-			return true
+		if indexCol != nil && col.Equal(nil, indexCol) {
+			isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
+			if isFullLen {
+				return true
+			}
+			if constVal == nil {
```

Comment:
The `constVal` is passed in as a parameter. We can move this check out of the for loop?